### PR TITLE
GUIs: XSpecGUI

### PR DESCRIPTION
**Pull request by profxj**
As named.
docs
notebook
no tests

### CODE REVIEW

**Comment by nhmc** [comment]
I'd like to comment on this before it's merged, but I won't be able to for
the next couple of days. Can it wait until then?

**Comment by ntejos** [comment]
Sure, no problem; )

**Comment by profxj** [comment]
No problem

On January 3, 2016 12:45:36 PM PST, Neil Crighton <notifications@github.com> wrote:
>I'd like to comment on this before it's merged, but I won't be able to
>for
>the next couple of days. Can it wait until then?
>
>On Mon, Jan 4, 2016 at 12:42 AM, J. Xavier Prochaska <
>notifications@github.com> wrote:
>
>> As named.
>> docs
>> notebook
>> no tests
>> ------------------------------
>> You can view, comment on, or merge this pull request online at:
>>
>>   https://github.com/linetools/linetools/pull/137
>> Commit Summary
>>
>>    - GUIs begin
>>    - GUIs begin
>>    - GUIs continue
>>    - doc
>>
>> File Changes
>>
>>    - *A* docs/examples/xspecgui.ipynb
>>    <https://github.com/linetools/linetools/pull/137/files#diff-0>
>(261)
>>    - *A* docs/guis.rst
>>    <https://github.com/linetools/linetools/pull/137/files#diff-1>
>(33)
>>    - *M* docs/install.rst
>>    <https://github.com/linetools/linetools/pull/137/files#diff-2> (3)
>>    - *M* docs/scripts.rst
>>    <https://github.com/linetools/linetools/pull/137/files#diff-3>
>(11)
>>    - *A* linetools/guis/__init__.py
>>    <https://github.com/linetools/linetools/pull/137/files#diff-4> (0)
>>    - *A* linetools/guis/line_widgets.py
>>    <https://github.com/linetools/linetools/pull/137/files#diff-5>
>(184)
>>    - *A* linetools/guis/simple.py
>>    <https://github.com/linetools/linetools/pull/137/files#diff-6> (6)
>>    - *A* linetools/guis/spec_widgets.py
>>    <https://github.com/linetools/linetools/pull/137/files#diff-7>
>(491)
>>    - *A* linetools/guis/utils.py
>>    <https://github.com/linetools/linetools/pull/137/files#diff-8>
>(213)
>>    - *A* linetools/guis/xspecgui.py
>>    <https://github.com/linetools/linetools/pull/137/files#diff-9>
>(115)
>>    - *A* linetools/scripts/lt_xspec.py
>>    <https://github.com/linetools/linetools/pull/137/files#diff-10>
>(48)
>>    - *M* setup.py
>>    <https://github.com/linetools/linetools/pull/137/files#diff-11>
>(1)
>>
>> Patch Links:
>>
>>    - https://github.com/linetools/linetools/pull/137.patch
>>    - https://github.com/linetools/linetools/pull/137.diff
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/linetools/linetools/pull/137>.
>>
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/linetools/linetools/pull/137#issuecomment-168539765
>
>!DSPAM:826,5689887325836165881379!

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: linetools/guis/spec_widgets.py [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+        # Analysis dict
+        self.adict['flg'] = 0  # Column density flag
+
+    def on_key(self, event):
+        """ Deals with key events
+        Parameters
+        ----------
+        event : event object
+        """
+        flg = -1
```
Comment (nhmc) [inline linetools/guis/spec_widgets.py]: Please add comment here saying what flg is used for.

File: linetools/scripts/lt_xspec.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+"""
+Plot a spectrum with an interactive QT GUI
+"""
+import pdb
+import sys
+
+from PyQt4 import QtGui
+
+from linetools.guis.xspecgui import XSpecGui
+
```
Comment (nhmc) [inline linetools/scripts/lt_xspec.py]: Please put the PyQt4 and xspecgui imports inside main(), to avoid imports before parsing the arguments.

**Comment by nhmc** [comment]
I left some inline comments. 

The velocity plot option is commented out, right? If so, please remove that option from the docs. Also in the docs, instead of `reload()`, use `import imp; imp.reload()` for python 3 compatibility. And fix the typo: "These key strokes help you explore the documenation" (should be explore the spectrum?)

File: linetools/guis/spec_widgets.py (hunk 1 above)
Comment (profxj) [inline linetools/guis/spec_widgets.py]: Done.

File: linetools/scripts/lt_xspec.py (hunk 2 above)
Comment (profxj) [inline linetools/scripts/lt_xspec.py]: Done.

**Comment by profxj** [comment]
Good catches on the docs.  Am going to implement the VelocityPlot soon, 
so I left that in with a 'coming soon' comment.

File: linetools/guis/line_widgets.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    """ Widget to set up spectral lines for plotting
+    """
+    def __init__(self, parent=None, status=None, init_llist=None, init_z=None):
+        """
+        Parameters
+        ----------
+        parent
+        status
+        init_llist
+        init_z
```
Comment (ntejos) [inline linetools/guis/line_widgets.py]: Please add basic documentation for these parameters.

File: linetools/guis/line_widgets.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        # Create a dialog window for redshift
+        z_label = QtGui.QLabel('z=')
+        self.zbox = QtGui.QLineEdit()
+        self.zbox.z_frmt = '{:.7f}'
+        self.zbox.setText(self.zbox.z_frmt.format(init_z))
+        self.zbox.setMinimumWidth(50)
+        self.connect(self.zbox, QtCore.SIGNAL('editingFinished ()'), self.setz)
+
+        # Create the line list
+        self.lists = ['None', 'ISM', 'Strong', 'Galaxy', 'H2', 'EUV', 'OVI']
```
Comment (ntejos) [inline linetools/guis/line_widgets.py]: Not convinced by having a 'OVI' linelist. Maybe worth considering having a 'Custom' linelist that can be given as a parameter in the command line (?)

File: linetools/guis/line_widgets.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+        self.llist_widget.currentItemChanged.connect(self.on_list_change)
+        self.llist_widget.setMaximumHeight(100)
+
+        # Input line list?
+        if init_llist is None:
+            self.llist = {} # Dict for the line lists
+            self.llist['Plot'] = False
+            self.llist['z'] = 0.
+            self.llist['List'] = 'None'
+            self.llist['Lists'] = []
```
Comment (ntejos) [inline linetools/guis/line_widgets.py]: Can you add a inline comment to explain the difference between self.llist['List'] and self.llist['Lists'] ?

File: linetools/guis/line_widgets.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        else: # Fill it all up and select
+            self.llist = init_llist
+            if not init_llist['List'] in self.lists:
+                self.lists.append(init_llist['List'])
+                self.llist_widget.addItem(init_llist['List'])
+                self.llist_widget.setCurrentRow(len(self.lists)-1)
+            else:
+                idx = self.lists.index(init_llist['List'])
+                self.llist_widget.setCurrentRow(idx)
+            try:
```
Comment (ntejos) [inline linetools/guis/line_widgets.py]: Please have the 'select' part of the code (i.e. current lines 75 and 76) in an independent block, after the `self.linelist` is fully initialized. You can then remove current line 73.

File: linetools/guis/line_widgets.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+                self.zbox.setText(self.zbox.z_frmt.format(init_llist['z']))
```
Comment (ntejos) [inline linetools/guis/line_widgets.py]: Have the `self.zbox` update in a different block too.

File: linetools/guis/line_widgets.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+        list_label = QtGui.QLabel('Line Lists:')
+        self.llist_widget = QtGui.QListWidget(self)
+        for ilist in self.lists:
+            self.llist_widget.addItem(ilist)
+        self.llist_widget.setCurrentRow(0)
+        self.llist_widget.currentItemChanged.connect(self.on_list_change)
+        self.llist_widget.setMaximumHeight(100)
+
+        # Input line list?
+        if init_llist is None:
```
Comment (ntejos) [inline linetools/guis/line_widgets.py]: Why not using `ltgu.set_llist()` to initialize the `self.llist`?

File: linetools/guis/spec_widgets.py [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+        """
+        # Flag to control re-draw
+        flg = -1
+
+        # NAVIGATING
+        if event.key in self.psdict['nav']:
+            flg = ltgu.navigate(self.psdict, event)
+
+        # DOUBLETS
+        if event.key in ['C', 'M', 'X', '4', '8', 'B']:  # Set left
```
Comment (ntejos) [inline linetools/guis/spec_widgets.py]: remove misleading comment (?)

File: linetools/guis/simple.py [hunk 10]
```diff
@@ -0,0 +1,6 @@
+""" Module for simple widgets
```
Comment (ntejos) [inline linetools/guis/simple.py]: Remove this file (?); do not think is being used.

**Comment by ntejos** [comment]
As a general comment, could you please add minimum docstring to the classes and methods (specially for those whose parameter types are not obvious from its variable name)?

After this and the other minor comments are addressed/considered I think it is ok to merge and work from there to improve the code. One obvious thing is to make 'navigation' keystrokes self consistent within linetools, and have these defined in a single file. I will raise an issue after xspecgui is merged.

File: linetools/guis/line_widgets.py (hunk 3 above)
Comment (profxj) [inline linetools/guis/line_widgets.py]: Done.

File: linetools/guis/line_widgets.py (hunk 4 above)
Comment (profxj) [inline linetools/guis/line_widgets.py]: Agreed.  Gone.

File: linetools/guis/line_widgets.py (hunk 5 above)
Comment (profxj) [inline linetools/guis/line_widgets.py]: Done.

File: linetools/guis/line_widgets.py (hunk 6 above)
Comment (profxj) [inline linetools/guis/line_widgets.py]: I don't see it.  Not changing.

File: linetools/guis/line_widgets.py (hunk 7 above)
Comment (profxj) [inline linetools/guis/line_widgets.py]: Same.  You can edit in a future PR.

File: linetools/guis/line_widgets.py (hunk 8 above)
Comment (profxj) [inline linetools/guis/line_widgets.py]: Not sure.

File: linetools/guis/spec_widgets.py (hunk 9 above)
Comment (profxj) [inline linetools/guis/spec_widgets.py]: Done.

File: linetools/guis/simple.py (hunk 10 above)
Comment (profxj) [inline linetools/guis/simple.py]: Removed.

**Comment by profxj** [comment]
Merging..

**Pull request merged**